### ISSUE TITLE
Generated class does not pass inits to embedded class in Constructor. NPE due to Path Initialization Problems.

### ISSUE THREAD

**Issue by tpischke** [querydsl/querydsl#1803]
The HibernateDomainExporter and the JPAAnnotationProcessor seem to produce different results.  Moreover, the JPAAnnotationProcessor produces an unusable result with PathInit Problems.

Functional Result using JPA Mapping File and HibernateDomainExporter:

```
public QMyEntity(Class<? extends MyEntity> type, PathMetadata metadata, PathInits inits) extends MyMappedSuperclass {
     super(type, metadata, inits);
     this._super = new QMyMappedSuperclass(type, metadata, inits);
     this.embeddedAttribute= inits.isInitialized("embeddedAttribute") ? new QMyEmbeddedAttribute(forProperty("embeddedAttribute"), inits.get("embeddedAttribute")) : null;
}
```

Nonfunctional result with Path Init Problems using JPAAnnotationGenerator and JPA Annotations.  QueryInit Annotation does not help because Constructor ignores them.  The second  **inits.get("embeddedAttribute")** argument to the Constructor of QMyEmbeddedAttribute is missing in the following generated code:

```
public QMyEntity(Class<? extends MyEntity> type, PathMetadata metadata, PathInits inits) extends MyMappedSuperclass {
        super(type, metadata, inits);
        this._super = new QMyMappedSuperclass(type, metadata, inits);
        this.embeddedAttribute = inits.isInitialized("embeddedAttribute") ? new QMyEmbeddedAttribue(forProperty("embeddedAttribute")) : null;
}
```

**Comment by tpischke** [comment]
I see that the HibernateDomainExporter has some logic for embedded properties that is missing from the JPADomainExporter.  Maybe that could be the reason.  Code looks pretty similar otherwise, same abstract class...

**Comment by tpischke** [comment]
Or maybe the difference lies in the difference between processing mapping files vs processing annotations.  We are in the process of converting from mapping files to annotations, and are blocked by this issue.

**Comment by tpischke** [comment]
If I manually correct the generated code to include the `inits.get("effectivity")` constructor argument, then everything works.  Seems like it should be a pretty simple fix

**Comment by tpischke** [comment]
It's probably worth mentioning that the embeddedAttribute is actually in a MappedSuperclass

**Comment by tpischke** [comment]
```
@Entity
public class MyEntityextends OtherClass {
    @Embedded
    @QueryInit("*")
    private EmbeddedAttribute embeddedAttribute;
}
```

**Comment by tpischke** [comment]
```
@Embeddable
public class EmbeddedAttribute {
    @ManyToOne(targetEntity = SomeEntity.class, fetch = FetchType.EAGER, cascade = CascadeType.REFRESH)
    @JoinColumn(name = "JOINCOLUMN_FK_ID", nullable = false)
    private SomeEntity attributeWithInitProblem;
```

**Comment by Shredder121** [comment]
@tpischke Thanks for the report and extensive writeup.
We're all not available right now, but tonight we can look into it.

**Comment by tpischke** [comment]
QueryDSL 4.0.8  
apt-maven-plugin 1.1.3

**Comment by tpischke** [comment]
```
 <plugin>
                    <groupId>com.mysema.maven</groupId>
                    <artifactId>apt-maven-plugin</artifactId>
                    <version>1.1.3</version>
                    <executions>
                        <execution>
                            <goals>
                                <goal>process</goal>
                            </goals>
                            <configuration>
                                <outputDirectory>target/generated-sources</outputDirectory>
                                <processor>com.querydsl.apt.jpa.JPAAnnotationProcessor</processor>      
                            </configuration>
                        </execution>
                    </executions>
                    <dependencies>
                        <dependency>
                            <groupId>com.querydsl</groupId>
                            <artifactId>querydsl-apt</artifactId>
                            <version>${querydsl.version}</version>
                        </dependency>
                    </dependencies>
                </plugin>
```

**Comment by tpischke** [comment]
It's a blocker for us, so hoping for a quick fix :)

**Comment by timowest** [comment]
I added a test case as a PR. Couldn't replicate it yet though.

**Comment by tpischke** [comment]
Great, let me know if I can help...

**Comment by timowest** [comment]
Could you provide a self contained test project that fails or maybe modify the PR that it fails / reproduces the problem?

**Comment by tpischke** [comment]
Probably, but it will take some time to isolate.  Won't get to it until Monday.

**Comment by tpischke** [comment]
Was hoping it would be immediately obvious to you. :)

**Comment by timowest** [comment]
@tpischke Could you confirm that you could replicate it with MyEntity and MyEmbeddedAttribute?

**Comment by tpischke** [comment]
No, was just documenting the basic situation with modified names. Actual classes are much more complicated.  If you can't reproduce I will try to find the minimal test case on Monday.

**Comment by tpischke** [comment]
A little more info:  The Problem is not reproducible if all involved classes are in a single maven module.  In the problem case, the EmbeddedAttribute class is in a separate jar from MyEntity.

**Comment by tpischke** [comment]
Perhaps I need to use the querydsl-maven-plugin for this use case?  I don't currently add generated code from the external jar to version control.  It compiles since I build all maven modules within a parent project, so the external jar gets compiled by maven before the downstream module.  It seems that the apt-maven-plugin gets a bit confused when generating the downstream classes however.  They compile, but have incorrect init logic.

I will do some testing with the querydsl-maven-plugin.

**Comment by tpischke** [comment]
When I use querydsl-maven-plugin, then I have different problems.  Generic Types no longer get handled properly.

**Comment by tpischke** [comment]
querydsl-maven-plugin seems to generate unwanted accesors for protected methods, and not just annotated fields.  apt-maven-plugin does not do this.  The accessors for the actual annotated fields are also incorrect with respect to generics.

**Comment by tpischke** [comment]
´´´
    @OneToMany(orphanRemoval = true, fetch = FetchType.EAGER, cascade = CascadeType.ALL, targetEntity = EigentumAnteilVersion.class)
    @JoinColumn(name = "MASTER_FK_ID", nullable = false)
    @OrderBy("effectivity.von DESC")
    private Set<EigentumAnteilVersion> versions = new TreeSet<>();
´´´

Gives:

´´´
    public final SetPath<ch.bedag.raum.shared.Version<?>, ch.bedag.raum.shared.QVersion> versions = this.<ch.bedag.raum.shared.Version<?>, ch.bedag.raum.shared.QVersion>createSet("versions", ch.bedag.raum.shared.Version.class, ch.bedag.raum.shared.QVersion.class, PathInits.DIRECT2);
´´

**Comment by tpischke** [comment]
I am unable to find a working generation configuration.  querydsl-maven-plugin and apt-maven-plugin exhibit different problems, but neither provides a workable solution.

**Comment by timowest** [comment]
@tpischke Could you provide a self-contained example domain that breaks?

**Comment by tpischke** [comment]
I've made a slightly simplified self-contained project that produces the problem with the inits using apt-maven-plugin.  I cannot upload the code to the issue tracker as it contains proprietary source.  How can I send it to you securely?

**Comment by tpischke** [comment]
I've sent some code to Timo's email address that should be sufficient to reproduce the problem.

**Comment by tpischke** [comment]
I see you've commited a fix in 1805. Does this commit fix the init problem?

**Comment by Shredder121** [comment]
> I've sent some code to Timo's email address that should be sufficient to reproduce the problem.

That's good, then Timo can at least look at it.

> I see you've commited a fix in 1805. Does this commit fix the init problem?

I'm not quite sure, it doesn't look like it should fix it, I think it's related to another issue.
But we'll look at yours soon then, thanks for taking the time to make a reproduction project.

**Comment by tpischke** [comment]
@Shredder121 Should I send you the code sample as well?  Haven't heard back from Timo yet...

**Comment by Shredder121** [comment]
I heard he was busy at work, but I'm sure he'll look at it soon.
Sorry for the delay.

**Comment by timowest** [comment]
Concerning the inits, I'll try to simplify the initialization model.

**Comment by tpischke** [comment]
Have you been able to reproduce the problem locally?  Still hoping for a fix, since this is blocking us from moving to Annotations.

**Comment by timowest** [comment]
I am working on it

**Comment by tpischke** [comment]
Glad to hear you're working on it!  Does that mean you were able to reproduce it?  I could maybe work on distilling the code I sent to something simpler, or adding an explicit unit test that fails if that would help.  I'm eager to insist wherever I can to help get this fixed.

**Comment by timowest** [comment]
I had some computer issues for a few days and wasn't able to advance this. I haven't yet found the culprit.

**Comment by tpischke** [comment]
Ok, thanks for the update.  Let me know if I can help in any way.

**Comment by tpischke** [comment]
Hi, I hate to be that annoying guy, but this hasn't moved in a while.  I'm getting a little nervous, since this is blocking our migration to JPA Annotations.  Should I start looking for possible alternatives and workarounds?

**Comment by timowest** [comment]
I haven't yet found a solution, but a possible workaround is to use JPADomainExporter, which exports JPA metamodels (works with annotated types and XML based configuration)

**Comment by tpischke** [comment]
Sure, I can try it.  How do I configure it?

**Comment by timowest** [comment]
Your code generation options based on JPA annotations are
* APT via JPAAnnotationProcessor (broken in your case)
* Reflection based using GenericExporter 
 * http://www.querydsl.com/static/querydsl/4.0.9/reference/html_single/#d0e2502
* JPA metamodel based using JPADomainExporter (doesn't require annotations)

JPADomainExporter example

```
EntityManagerFactory emf = Persistence.createEntityManagerFactory("h2", new Properties());
JPADomainExporter exporter = new JPADomainExporter(outputFolder, emf.getMetamodel());
exporter.execute();
```

**Comment by timowest** [comment]
I'd start with the GenericExporter based approach

**Comment by mnshdw** [comment]
Hello,

I have the same problem as @tpischke, and attached a sample maven project to reproduce it. If you look at QAccountUserAccess in my example, it is missing **inits(id)** which leads to a NullPointerException when querying on it.

[querydsl-i1803.tar.gz](https://github.com/querydsl/querydsl/files/295226/querydsl-i1803.tar.gz)

**Comment by Shredder121** [comment]
@mnshdw Thanks for the project, could you also add a test that we can run on the project?
Maybe put it up on GitHub? (Not really necessary though)

**Comment by mnshdw** [comment]
Hi @Shredder121,
Here you go : https://github.com/mnshdw/querydsl-i1803
